include/faabric/scheduler/FunctionMigrationThread.h

```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+namespace faabric::scheduler {
+class FunctionMigrationThread
+{
+  public:
+    // Start a background thread that, every wake up period, will check if there
+    // are migration opportunities for in-flight apps that have opted in to
+    // being checked for migrations.
+    void start(int wakeUpPeriodSecondsIn);
+
+    void stop();
+
+    int wakeUpPeriodSeconds;
+
+  private:
+    std::unique_ptr<std::thread> workThread = nullptr;
+    std::mutex mx;
+    std::condition_variable mustStopCv;
```

Comment:
Now that we have moved to c++20, new code could use `std::jthread` with a stop_token instead of this relatively complex pattern, see:
```c++
#include <thread>
#include <iostream>
 
using namespace std::literals::chrono_literals;
 
void f(std::stop_token stop_token, int value)
{
    while (!stop_token.stop_requested()) {
        std::cout << value++ << ' ' << std::flush;
        std::this_thread::sleep_for(200ms);
    }
    std::cout << std::endl;
}
 
int main()
{
    std::jthread thread(f, 5); // prints 5 6 7 8... for approximately 3 seconds
    std::this_thread::sleep_for(3s);
    // The destructor of jthread calls request_stop() and join().
}
```
from <https://en.cppreference.com/w/cpp/thread/stop_token>

Reply:
Nice, didn't know about this. 

Rather than have new and old code diverge in their styles, this should be straightforward to do wholesale in an isolated PR. Have created an issue: https://github.com/faasm/faabric/issues/216.

Reply:
(I realise I'm letting the great be the enemy of the good here, but we're in a rush at the moment and as this is already working I'm reluctant to change it, even though the result would be simpler and cleaner)

Reply:
Let's try and capture as many instances of background thread logic as possible in a separate PR (including the bg threads in `FaabricMain`).